### PR TITLE
[KAFKA-38] Read records implementation 

### PR DESCRIPTION
**Pull request by paulgrape**
"Signed-off-by: [Pavel Vinogradov] [vin.pavel13@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

### CODE REVIEW

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 1]
```diff
@@ -73,7 +135,24 @@ private void submitRecordOperation(DataOperation op, OpType opType) {
...
+      val adminConfig =
+          adminClientCreateFuncCache.computeIfAbsent(config, AdminClientCreateFunctionImpl::new);
+      val adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
+
+      val consConfig = consumerConfigCache.computeIfAbsent(nodeAddr, this::createConsumerConfig);
+      val consumerConfig =
+          consumerCreateFuncCache.computeIfAbsent(consConfig, ConsumerCreateFunctionImpl::new);
+      val kafkaConsumer = consumerCache.computeIfAbsent(nodeAddr, consumerConfig);
+
+      val result = kafkaConsumer.poll(Duration.ofSeconds(10));
```
Comment (paulgrape) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: How should I process received records?

Comment (akurilov) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: happy path:
1. extract the record from the result
2. extract the record payload (value) size
3. call `countBytesDone` method for the operation instance
4. set the `SUCC` status for the operations instance

**Review by akurilov — CHANGES_REQUESTED** [review]
Please resolve the merge conflicts w/ master

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 1 above)
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: ths timeout should be a configuration option, e.g.
`storage-driver-read-timeoutMillis`

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 2]
```diff
@@ -108,6 +187,33 @@ private void submitNoop(final O op) {
     completeOperation(op, SUCC);
   }
 
+  private Properties createConfig(String nodeAddr) {
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: please make such methods package-private to make the code more testable

**Review by akurilov — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 1 above)
Comment (paulgrape) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Done

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 2 above)
Comment (paulgrape) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Done

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 3]
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
 
   private void submitRecordDeleteOperation() {}
 
-  private void submitRecordReadOperation() {}
+  private void submitRecordReadOperation(final DataOperation recordOp, String nodeAddr) {
+    try {
+      val config = configCache.computeIfAbsent(nodeAddr, this::createConfig);
+      val adminConfig =
+          adminClientCreateFuncCache.computeIfAbsent(config, AdminClientCreateFunctionImpl::new);
+      val adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: unused

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 4]
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
...
+
+      val consConfig = consumerConfigCache.computeIfAbsent(nodeAddr, this::createConsumerConfig);
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: consumerConfig

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 5]
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
...
+      val consumerConfig =
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:140]: consumerCreateFunc

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java [hunk 6]
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
...
+          consumerCreateFuncCache.computeIfAbsent(consConfig, ConsumerCreateFunctionImpl::new);
+      val kafkaConsumer = consumerCache.computeIfAbsent(nodeAddr, consumerConfig);
+
+      val result = kafkaConsumer.poll(Duration.ofMillis(readTimeout));
```
Comment (akurilov) [inline src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:147]: please invoke the `startRequest`, `finishRequest` methods before this and
`startResponse`, `finishResponse` after

**Review by akurilov — CHANGES_REQUESTED** [review]

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 6 above)
Comment (paulgrape) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:147]: `finishRequest`, `startResponse` and `finishResponse` are called in `completeOperation` method

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 3 above)
Comment (paulgrape) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java]: Removed

File: src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java (hunk 6 above)
Comment (akurilov) [reply src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:147]: ok, let's change this later

**Review by akurilov — DISMISSED** [review]

**Review by akurilov — APPROVED** [review]

**Pull request merged**